tt_eager/tensor/tensor.cpp

```diff
@@ -116,6 +142,94 @@ void Tensor::deallocate(bool force) {
     }
 }
 
+bool Tensor::metadata_populated() const {
+    // Make a list of boolens for storage
+    return this->tensor_attributes->metadata_populated;
+}
+
+void Tensor::deepcopy(const Tensor& other) {
+    // Wait until the tensor being copied is populated
+    while (not other.metadata_populated()) {}
+    // Populate tensor metadata
+    this->set_shape(other.get_shape());
+    this->set_storage(other.get_storage());
+    this->set_dtype(other.get_dtype());
+    this->set_layout(other.get_layout());
+    // Set metadata populated flag for getters
+    this->tensor_attributes->metadata_populated = true;
+}
+
+void Tensor::populate_buffers_and_metadata(const Tensor& other) {
+    // Similar to deepcopy, but to be applied on a tensor that has an empty storage
+    // container initialized. Require tensor storage to be correctly initialized.
+    // Populate storage container with buffers + shapes
+    std::visit([this] (auto&& storage) {
+        using StorageType = std::decay_t<decltype(storage)>;
+        if constexpr(std::is_same_v<StorageType, OwnedStorage> or std::is_same_v<StorageType, DeviceStorage>) {
+            std::get<StorageType>(this->tensor_attributes->storage).buffer = storage.buffer;
+        } else if constexpr(std::is_same_v<StorageType, MultiDeviceHostStorage> or std::is_same_v<StorageType, MultiDeviceStorage>) {
+            std::get<StorageType>(this->tensor_attributes->storage).buffers = storage.buffers;
+            std::get<StorageType>(this->tensor_attributes->storage).shapes = storage.shapes;
+        }
+    }, other.get_storage()); // Non blocking storage query, since this is done for tensors that get created inside the worker thread
+    // Populate remaining MD
+    this->set_shape(other.get_shape());
+    this->set_dtype(other.get_dtype());
+    this->set_layout(other.get_layout());
+    this->tensor_attributes->metadata_populated = true;
+}
+
+std::vector<Device*> Tensor::get_workers(bool blocking) const {
+    bool wait_for_tensor_populated = (Device::get_worker_mode() == Device::WorkerQueueMode::SYNCHRONOUS) or blocking;
+    TT_FATAL(wait_for_tensor_populated or (this->workers.size()), "Worker Handles for tensor must be populated or blocking = true must be set in get_workers().");
+    if (not this->workers.size()) {
+        if (wait_for_tensor_populated) {
+            while (not this->metadata_populated());
+            std::vector<Device*> workers = {};
+            std::visit(
+            [&] (auto&& storage) {
+                using StorageType = std::decay_t<decltype(storage)>;
+                if constexpr (std::is_same_v<StorageType, DeviceStorage>) {
+                    workers = {this->device()};
+                }
+                else if constexpr (std::is_same_v<StorageType, MultiDeviceStorage>) {
+                    workers.reserve(storage.buffers.size());
+                    for (const auto& buffer : storage.buffers) {
+                        workers.push_back(buffer->device());
```

Comment:
why do we have to populate the workers here, rather than a spin loop like metadata_populated()? ie. the following

```
while (not this->workers_populated());
return this->workers;
```

Reply:
This was to handle the degen case where a user chooses not to set the workers list through the tensor constructor. Rather than erroring out, I thought it might be nicer to just handle it and infer the workers from the devices.